controller/src/main/java/io/pravega/controller/store/stream/Cache.java

```diff
@@ -15,46 +15,50 @@
 
 import javax.annotation.ParametersAreNonnullByDefault;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
 import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
 
+/**
+ * Cache for asynchronously retrieving and loading records from underlying store using the supplied loader.
+ * This uses Guava's loading cache and takes a loader function for loading entries into the cache. 
+ * This class caches Futures which hold the metadata record with version. The cache is untyped and the CompletableFutures
+ * can hold any value under the VersionedMetadata wrapper.
+ * The values are by default held for 2 minutes after creation unless invalidated explicitly.
+ * The maximum number of records that can be held in the cache is 10000. 
+ */
 public class Cache {
+    private static final int MAX_CACHE_SIZE = 10000;
+    
+    private final LoadingCache<CacheKey, CompletableFuture<VersionedMetadata<?>>> cache;
 
-    @FunctionalInterface
-    public interface Loader {
-        CompletableFuture<Data> get(final String key);
-    }
-
-    private final LoadingCache<String, CompletableFuture<Data>> cache;
-
-    public Cache(final Loader loader) {
+    public Cache(final Function<CacheKey, CompletableFuture<VersionedMetadata<?>>> loader) {
         cache = CacheBuilder.newBuilder()
-                .maximumSize(10000)
-                .expireAfterWrite(10, TimeUnit.MINUTES)
-                .build(new CacheLoader<String, CompletableFuture<Data>>() {
-                    @ParametersAreNonnullByDefault
-                    @Override
-                    public CompletableFuture<Data> load(final String key) {
-                        CompletableFuture<Data> result = loader.get(key);
-                        result.exceptionally(ex -> {
-                            invalidateCache(key);
-                            return null;
-                        });
-                        return result;
-                    }
-                });
+                            .maximumSize(MAX_CACHE_SIZE)
+                            .expireAfterAccess(2, TimeUnit.MINUTES)
```

Comment:
This is looking like a magic number, we don't know whether 2 minutes or 10 minutes or anything is a good value, is it right?

Reply:
Yeah its a magic number.. Typically if value is regularly accessed, it should be in the cache, or LRU will be removed if the cache is full.